src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -241,12 +241,16 @@ public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancella
                 {
                     ClearCancellationToken();
 
-                    if (cancellationToken.IsCancellationRequested)
+                    if (tokenSource.IsCancellationRequested && !cancellationToken.IsCancellationRequested)
+                    {
+                        // Catch cancellation and translate it into setting isCanceled = true
+                        isCanceled = true;
+                    }
+                    else
                     {
                         throw;
                     }
```

Comment:
Nit: just a style consistency thing, but here it's using:
```C#
else
{
    throw;
}
```
and below it's just using:
```C#
throw;
```